app/models/oai_provider/valkyrie_provider_model.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+class OaiProvider < OAI::Provider::Base
+  class ValkyrieProviderModel < OAI::Provider::Model
+    require_relative "oai_wrapper"
+    def earliest
+      Time.zone.at(0)
+    end
+
+    def latest
+      Time.current
+    end
+
+    def sets
+      @sets ||= change_set_persister.query_service.find_all_of_model(model: Collection).map { |coll| OAI::Set.new(name: coll.title.first, spec: coll.slug.first) }
+    end
+
+    def deleted?
+      false
+    end
+
+    # find is the core method of a model, it returns records from the model
+    # bases on the parameters passed in.
+    #
+    # <tt>selector</tt> can be a singular id, or the symbol :all
+    # <tt>options</tt> is a hash of options to be used to constrain the query.
+    #
+    # Valid options:
+    # * :from => earliest timestamp to be included in the results
+    # * :until => latest timestamp to be included in the results
+    # * :set => the set from which to retrieve the results
+    # * :metadata_prefix => type of metadata requested (this may be useful if
+    #                       not all records are available in all formats)
+    def find(selector, options = {})
+      return next_set(options) if options[:resumption_token]
+      if selector == :all
+        find_all(options)
+      else
+        OAIWrapper.new(change_set_persister.query_service.find_by(id: selector))
+      end
+    end
+
+    class Options < DelegateClass(Hash)
+      def marc?
+        fetch(:metadata_prefix, "").casecmp("marc21").zero?
+      end
+
+      def next_offset
+        return 0 unless key?(:last)
+        fetch(:last) + 10
+      end
+
+      def set
+        fetch(:set, nil)
+      end
+    end
+
+    def find_all(options)
+      return next_set(options[:resumption_token]) if options[:resumption_token]
+      options = Options.new(options)
+      result = change_set_persister.query_service.custom_queries.paged_all(
+        offset: options.next_offset,
+        only_models: ScannedResource,
+        collection_slug: options.set,
+        marc_only: options.marc?
+      ).map { |r| OAIWrapper.new(r) }.to_a
+      wrap_results(result, options)
+    end
+
+    def wrap_results(result, options)
+      return result if result.length < 10
+      OAI::Provider::PartialResult.new(
+        result,
+        OAI::Provider::ResumptionToken.new(options)
+      )
+    end
+
+    def next_set(token_string)
+      token = OAI::Provider::ResumptionToken.parse(token_string)
+      options = { last: 0 }.merge(token.to_conditions_hash)
+      find_all(options)
+    end
+
+    def change_set_persister
```

Comment:
It's sort of confusing for this to be in here when it looks like you're only using the query service